tests/subsys/storage/flash_buffered_write/src/main.c

```diff
@@ -0,0 +1,302 @@
+
+	init_target();
+
+	/* Test when write spans multiple pages crosses border of page */
+	rc = fbw_write(&ctx, write_buf, (page_size * num_pages) + 128, false);
+	zassert_equal(rc, 0, "expected success");
+
+	/* First three pages should be written */
+	VERIFY_WRITTEN(0, page_size * num_pages);
+
+	/* Fill rest of the page */
+	rc = fbw_write(&ctx, write_buf, page_size - 128, false);
+	zassert_equal(rc, 0, "expected success");
+
+	/* First four pages should be written */
+	VERIFY_WRITTEN(0, BUF_LEN * (num_pages + 1));
+}
+
+static void test_fbw__bytes_written(void)
+{
+	int rc;
+	size_t offset;
+
+	init_target();
+
+	/* Verify that the offset is retained across failed downolads */
+	rc = fbw_write(&ctx, write_buf, BUF_LEN + 128, false);
+	zassert_equal(rc, 0, "expected success");
+
+	/* First page should be written */
+	VERIFY_WRITTEN(0, BUF_LEN);
+
+	/* Fill rest of the page */
+	offset = fbw_bytes_written(&ctx);
+	zassert_equal(offset, BUF_LEN, "offset should match buf size");
+
+	/* Fill up the buffer MINUS 128 to verify that write_buf_pos is kept */
+	rc = fbw_write(&ctx, write_buf, BUF_LEN - 128, false);
+	zassert_equal(rc, 0, "expected success");
+
+	/* Second page should be written */
+	VERIFY_WRITTEN(BUF_LEN, BUF_LEN);
+}
+
+static void test_fbw__write_callback(void)
+{
+	int rc;
+
+	init_target();
+
+	/* Trigger verification in callback */
+	cb_buf = buf;
+	cb_len = BUF_LEN;
+	cb_offset = FLASH_BASE;
+
+	rc = fbw_write(&ctx, write_buf, BUF_LEN + 128, false);
+	zassert_equal(rc, 0, "expected success");
+
+	cb_len = BUF_LEN;
+	cb_offset = FLASH_BASE + BUF_LEN;
+
+	/* Fill rest of the buffer */
+	rc = fbw_write(&ctx, write_buf, BUF_LEN - 128, false);
+	zassert_equal(rc, 0, "expected success");
+	VERIFY_WRITTEN(BUF_LEN, BUF_LEN);
+
+	/* Fill half of the buffer and flush it to flash */
+	cb_len = BUF_LEN/2;
+	cb_offset = FLASH_BASE + (2 * BUF_LEN);
+
+	rc = fbw_write(&ctx, write_buf, BUF_LEN/2, true);
+	zassert_equal(rc, 0, "expected success");
+
+	/* Ensure that failing callback trickles up to caller */
+	cb_ret = -EFAULT;
+	cb_buf = NULL; /* Don't verify other parameters of the callback */
+	rc = fbw_write(&ctx, write_buf, BUF_LEN, true);
+	zassert_equal(rc, -EFAULT, "expected failure from callback");
+}
+
+#ifdef CONFIG_FBW_ERASE
+static void test_fbw__erase(void)
+{
+	int rc;
+
+	init_target();
+
+	/* Write out one buf */
+	rc = fbw_write(&ctx, write_buf, BUF_LEN, false);
+	zassert_equal(rc, 0, "expected success");
+
+	rc = fbw_erase(&ctx, FLASH_BASE);
+	zassert_equal(rc, 0, "expected success");
+
+	VERIFY_ERASED(FLASH_BASE, page_size);
+}
+#endif
+
+void test_main(void)
+{
+	fdev = device_get_binding(FLASH_NAME);
+	api = fdev->driver_api;
+	api->page_layout(fdev, &layout, &layout_size);
+
+	page_size = layout->pages_size;
+	__ASSERT_NO_MSG(page_size > BUF_LEN);
+
+	ztest_test_suite(lib_fbw_test,
+	     ztest_unit_test(test_fbw__init),
+	     ztest_unit_test(test_fbw__write),
+	     ztest_unit_test(test_fbw__write_cross_buf_border),
```

Comment:
Just curious... what's up with the double underscores here? Seems weird.

Reply:
Just something I used to do back in the day to make it easier to read what tests were passing on an old CI system we had, will change it back to how its done in the other tests.